test-tls13-finished.py: failed with unexpected NewSessionTicket

# Bug Report
The tests `padding - cipher TLS_AES_256_GCM_SHA384, pad_byte 0, pad_left 0, pad_right 16777167` and `padding - cipher TLS_AES_128_GCM_SHA256, pad_byte 0, pad_left 0, pad_right 16777183` fail with NewSessionTicket being received.

That is, with gnutls implementing the sending of new session tickets early:
https://gitlab.com/gnutls/gnutls/merge_requests/711

Output:
```
test-tls13-finished.py:stdout:padding - cipher TLS_AES_128_GCM_SHA256, pad_byte 0, pad_left 0, pad_right 16777183 ...
server:stderr:|<2>| added 3 protocols, 29 ciphersuites, 18 sig algos and 8 groups into priority list
server:stderr:|<3>| ASSERT: buffers.c[get_last_packet]:1172
server:stderr:|<2>| checking 13.01 (GNUTLS_AES_128_GCM_SHA256) for compatibility
server:stderr:|<3>| ASSERT: server_name.c[gnutls_server_name_get]:235
server:stderr:|<2>| Selected (RSA-PSS) cert based on ciphersuite 13.1: GNUTLS_AES_128_GCM_SHA256
server:stderr:|<2>| EXT[0x1ec6460]: server generated SECP256R1 shared key
...
server:stderr:Scheduling inactive connection for close
test-tls13-finished.py:stdout:Error encountered while processing node <tlsfuzzer.expect.ExpectClose object at 0x7fbd073df1d0> (child: None) with last message being: <tlslite.messages.Message object at 0x7fbd07395410>
test-tls13-finished.py:stdout:Error while processing
test-tls13-finished.py:stdout:Traceback (most recent call last):
test-tls13-finished.py:stdout:  File "scripts/test-tls13-finished.py", line 421, in main
test-tls13-finished.py:stdout:    runner.run()
test-tls13-finished.py:stdout:  File "/home/nmavrogi/cvs/gnutls/tests/suite/tls-fuzzer/tlsfuzzer/tlsfuzzer/runner.py", line 210, in run
test-tls13-finished.py:stdout:    RecordHeader2)))
test-tls13-finished.py:stdout:AssertionError: Unexpected message from peer: Handshake(new_session_ticket)
test-tls13-finished.py:stdout:
test-tls13-finished.py:stdout:padding - cipher TLS_AES_256_GCM_SHA384, pad_byte 0, pad_left 0, pad_right 16777167 ...
```

(the other case is identical)

What I see in wireshark (attached) is the server's part of the handshake, and the transfer of the packets from the client until a tcp session close. That seems to be exactly what the current code is supposed to handle, though that doesn't happen, and I'm not sure why.

[capture.zip](https://github.com/tomato42/tlsfuzzer/files/2222655/capture.zip)

@nmav yes that gnutls#MR711 fixes issue of encryption, there's just minor issue of sending incorrect alert in case the Finished is really big:

```
padding - cipher TLS_AES_128_GCM_SHA256, pad_byte 0, pad_left 0, pad_right 16777183 ...
Error encountered while processing node <tlsfuzzer.expect.ExpectAlert object at 0x7fd723e9acd0> (child: <tlsfuzzer.expect.ExpectClose object at 0x7fd723e9ac90>) with last message being: <tlslite.messages.Message object at 0x7fd723b69c50>
Error while processing
Traceback (most recent call last):
  File "scripts/test-tls13-finished.py", line 435, in main
    runner.run()
  File "/home/hkario/dev/tlsfuzzer/tlsfuzzer/runner.py", line 212, in run
    node.process(self.state, msg)
  File "/home/hkario/dev/tlsfuzzer/tlsfuzzer/expect.py", line 1110, in process
    raise AssertionError(problem_desc)
AssertionError: Expected alert description "decode_error" does not match received "internal_error"
```

--- a/scripts/test-tls13-finished.py
+++ b/scripts/test-tls13-finished.py
@@ -17,7 +17,7 @@ from tlsfuzzer.expect import ExpectServerHello, ExpectCertificate, \
         ExpectServerHelloDone, ExpectChangeCipherSpec, ExpectFinished, \
         ExpectAlert, ExpectApplicationData, ExpectClose, \
         ExpectEncryptedExtensions, ExpectCertificateVerify, \
-        ExpectNewSessionTicket
+        ExpectNewSessionTicket, ExpectNoMessage
 
 from tlslite.constants import CipherSuite, AlertLevel, AlertDescription, \
         TLS_1_3_DRAFT, GroupName, ExtensionType, SignatureScheme
@@ -368,6 +368,20 @@ def main():
         node = node.add_child(ExpectCertificate())
         node = node.add_child(ExpectCertificateVerify())
         node = node.add_child(ExpectFinished())
+
+        # (the crazy handling of the messages below is because we are
+        # sending one message (Finished) in multiple records, and server
+        # can abort the connection after processing any number of records)
+
+        # conditionally wait for NewSessionTicket messages
+        # this will help in case the server does send early NST (without
+        # waiting for client Finished) but will abort reading of the
+        # Finished after one record
+        no_message = node.add_child(ExpectNoMessage(0.001))
+        nst = ExpectNewSessionTicket()
+        no_message.next_sibling = nst
+        nst.add_child(no_message)
+        node = no_message
 
         # alert+close can happen during sending large Finished message,
         # therefore we are specifying it as its sibling

--- a/scripts/test-tls13-finished.py
+++ b/scripts/test-tls13-finished.py
@@ -378,7 +378,7 @@ def main():
         # waiting for client Finished) but will abort reading of the
         # Finished after one record
         no_message = node.add_child(ExpectNoMessage(0.001))
-        nst = ExpectNewSessionTicket()
+        nst = ExpectNewSessionTicket(note='first')
         no_message.next_sibling = nst
         nst.add_child(no_message)
         node = no_message
@@ -395,7 +395,7 @@ def main():
         node.next_sibling = close_node
 
         # This message may be sent right after server finished
-        cycle = ExpectNewSessionTicket()
+        cycle = ExpectNewSessionTicket(note='second')
         node = node.add_child(cycle)
         node.add_child(cycle)
 

--- a/scripts/test-tls13-finished.py
+++ b/scripts/test-tls13-finished.py
@@ -29,7 +29,7 @@ from tlslite.extensions import KeyShareEntry, ClientKeyShareExtension, \
 from tlsfuzzer.helpers import key_share_gen
 
 
-version = 1
+version = 2
 
 
 def help_msg():
@@ -304,8 +304,10 @@ def main():
         (CipherSuite.TLS_AES_128_GCM_SHA256, 0, 0, 16), # SHA-384 size
         (CipherSuite.TLS_AES_128_GCM_SHA256, 0, 0, 32),
         (CipherSuite.TLS_AES_128_GCM_SHA256, 0, 0, 48),
-        (CipherSuite.TLS_AES_128_GCM_SHA256, 0, 0, 2**14-4-32),
-        (CipherSuite.TLS_AES_128_GCM_SHA256, 0, 0, 256**3-1-32),
+        (CipherSuite.TLS_AES_128_GCM_SHA256, 0, 0, 2**14-4-32), # max record
+        (CipherSuite.TLS_AES_128_GCM_SHA256, 0, 0, 0x20000), # intermediate
+        (CipherSuite.TLS_AES_128_GCM_SHA256, 0, 0, 0x30000), # bigger than max ClientHello
+        (CipherSuite.TLS_AES_128_GCM_SHA256, 0, 0, 256**3-1-32), # max handshake
         (CipherSuite.TLS_AES_128_GCM_SHA256, 0, 1, 0),
         (CipherSuite.TLS_AES_128_GCM_SHA256, 0, 2, 0),
         (CipherSuite.TLS_AES_128_GCM_SHA256, 0, 4, 0),
@@ -313,7 +315,7 @@ def main():
         (CipherSuite.TLS_AES_128_GCM_SHA256, 0, 16, 0), # SHA-384 size
         (CipherSuite.TLS_AES_128_GCM_SHA256, 0, 32, 0),
         (CipherSuite.TLS_AES_128_GCM_SHA256, 0, 48, 0),
-        (CipherSuite.TLS_AES_128_GCM_SHA256, 0, 2**14-4-32, 0),
+        (CipherSuite.TLS_AES_128_GCM_SHA256, 0, 2**14-4-32, 0), # max record
         (CipherSuite.TLS_AES_128_GCM_SHA256, 0, 12, 0),
         (CipherSuite.TLS_AES_128_GCM_SHA256, 0, 1, 1),
         (CipherSuite.TLS_AES_128_GCM_SHA256, 0, 8, 8), # SHA-384 size
@@ -324,8 +326,10 @@ def main():
         (CipherSuite.TLS_AES_256_GCM_SHA384, 0, 0, 16), # SHA-512 size
         (CipherSuite.TLS_AES_256_GCM_SHA384, 0, 0, 32),
         (CipherSuite.TLS_AES_256_GCM_SHA384, 0, 0, 48),
-        (CipherSuite.TLS_AES_256_GCM_SHA384, 0, 0, 2**14-4-48),
-        (CipherSuite.TLS_AES_256_GCM_SHA384, 0, 0, 256**3-1-48),
+        (CipherSuite.TLS_AES_256_GCM_SHA384, 0, 0, 2**14-4-48), # max record
+        (CipherSuite.TLS_AES_256_GCM_SHA384, 0, 0, 0x20000),
+        (CipherSuite.TLS_AES_256_GCM_SHA384, 0, 0, 0x30000), # bigger than max ClientHello
+        (CipherSuite.TLS_AES_256_GCM_SHA384, 0, 0, 256**3-1-48), # max handshake
         (CipherSuite.TLS_AES_256_GCM_SHA384, 0, 0, 12),
         (CipherSuite.TLS_AES_256_GCM_SHA384, 0, 1, 0),
         (CipherSuite.TLS_AES_256_GCM_SHA384, 0, 2, 0),
@@ -334,7 +338,7 @@ def main():
         (CipherSuite.TLS_AES_256_GCM_SHA384, 0, 16, 0), # SHA-512 size
         (CipherSuite.TLS_AES_256_GCM_SHA384, 0, 32, 0),
         (CipherSuite.TLS_AES_256_GCM_SHA384, 0, 48, 0),
-        (CipherSuite.TLS_AES_256_GCM_SHA384, 0, 2**14-4-48, 0),
+        (CipherSuite.TLS_AES_256_GCM_SHA384, 0, 2**14-4-48, 0), # max record
         (CipherSuite.TLS_AES_256_GCM_SHA384, 0, 1, 1),
         (CipherSuite.TLS_AES_256_GCM_SHA384, 0, 8, 8) # SHA-512 size
         ]

--- a/tests/test_tlsfuzzer_expect.py
+++ b/tests/test_tlsfuzzer_expect.py
@@ -1629,6 +1629,17 @@ class TestExpectNewSessionTicket(unittest.TestCase):
         self.assertIn(nst, state.session_tickets)
         self.assertIsNotNone(state.session_tickets[0].time)
 
+    def test___repr__(self):
+        exp = ExpectNewSessionTicket()
+
+        self.assertEqual("ExpectNewSessionTicket()", repr(exp))
+
+    def test___repr___with_note(self):
+        exp = ExpectNewSessionTicket(note="some string")
+
+        self.assertEqual("ExpectNewSessionTicket(note='some string')",
+                         repr(exp))
+
 
 class TestExpectVerify(unittest.TestCase):
     def test___init__(self):

--- a/tlsfuzzer/expect.py
+++ b/tlsfuzzer/expect.py
@@ -1031,12 +1031,27 @@ class ExpectEncryptedExtensions(ExpectHandshake):
 
 class ExpectNewSessionTicket(ExpectHandshake):
     """Processing TLS handshake protocol new session ticket message."""
-    def __init__(self):
+
+    def __init__(self, note=None):
+        """
+        Initialise object.
+
+        Note that "note" parameter MUST be specified as a keyword argument,
+        i.e. read the definition as `(self, *, note=None)` (see PEP 3102).
+        Otherwise the behaviour of this node is not guaranteed if new
+        arguments are added to it (as they will be added *before* the "note"
+        argument).
+
+        :param str note: name or comment attached to the node, will be printed
+           when str() or repr() is called on the node
+        """
         super(ExpectNewSessionTicket, self).__init__(
             ContentType.handshake,
             HandshakeType.new_session_ticket)
+        self.note = note
 
     def process(self, state, msg):
+        """Parse, verify and process the message."""
         assert msg.contentType == ContentType.handshake
         parser = Parser(msg.write())
         hs_type = parser.get(1)
@@ -1046,6 +1061,12 @@ class ExpectNewSessionTicket(ExpectHandshake):
         ticket.time = time.time()
 
         state.session_tickets.append(ticket)
+
+    def __repr__(self):
+        """Return human readable representation of object."""
+        return "ExpectNewSessionTicket({0})".format(
+            ", ".join("{0}={1!r}".format(name, getattr(self, name)) for name in
+                      ['note'] if getattr(self, name) is not None))
 
 
 class ExpectAlert(Expect):

